src/bigtable/mod.rs

```diff
@@ -0,0 +1,594 @@
+        }
+        None
+    }
+}
+
+/// A client for connecting to bigtable. Created from the
+/// [`build_bigtable_client`](crate::builder::ClientBuilder::build_bigtable_client)
+/// function.
+#[derive(Clone)]
+pub struct BigtableClient<
+    C = crate::DefaultConnector,
+    Retry = ExponentialBackoff<BigtableRetryCheck>,
+> {
+    inner: api::bigtable::v2::bigtable_client::BigtableClient<
+        AuthGrpcService<tonic::transport::Channel, OAuthTokenSource<C>>,
+    >,
+    retry: Retry,
+    table_prefix: String,
+}
+
+impl<C, Retry> BigtableClient<C, Retry>
+where
+    C: crate::Connect + Clone + Send + Sync + 'static,
+    Retry: RetryPolicy<(), tonic::Status> + 'static,
+    Retry::RetryOp: Send + 'static,
+    <Retry::RetryOp as RetryOperation<(), tonic::Status>>::Sleep: Send + 'static,
+{
+    /// Request some rows from bigtable.
+    ///
+    /// This is the most general read request; various other convenience methods are
+    /// available.
+    pub fn read_rows(
+        &mut self,
+        mut request: ReadRowsRequest,
+    ) -> impl Stream<Item = Result<Row, tonic::Status>> + '_ {
+        async_stream::stream! {
+            let mut retry = self.retry.new_operation();
+            let mut state = ReadInProgress::default();
+
+            // Keep track of the last returned (and/or last scanned) key, so that we won't
+            // request it again if we need to retry.
+            let mut last_key: Option<Bytes> = None;
+
+            let mut response = self.inner.read_rows(request.clone()).await?.into_inner();
+            let result = loop {
+                let message = match response.next().await {
+                    Some(m) => m,
+                    None => break Ok(()),
+                };
+
+                let message = match message {
+                    Ok(m) => m,
+                    Err(e) => break Err(e),
+                };
+
+                last_key = Some(
+                    last_key
+                        .unwrap_or_default()
+                        .max(message.last_scanned_row_key),
+                );
+
+                for chunk in message.chunks {
+                    if let Some(row) = state.process_chunk(chunk) {
+                        last_key = Some(last_key.unwrap_or_default().max(row.key.clone()));
+                        yield Ok(row);
+                    }
+                }
+            };
+
+            if let Err(e) = result {
+                if let Some(sleep) = retry.check_retry(&(), &e) {
+                    sleep.await;
+                } else {
+                    yield Err(e.into());
+                    return;
+                }
+            } else {
+                return;
+            }
+
+            // If we get to this point, there was a retry-able error. We'll loop back and retry the
+            // request again, but first let's modify the request to avoid re-requesting previously-returned
+            // data.
+            if let Some(key) = last_key {
```

Comment:
there isn't a loop around the `inner.read_rows` call, this will exit the stream without retrying. There need to be 2 loops i think: 1 around inner.read_rows and 1 around response.next.

As a matter of convention, I prefer using explicit `continue`s instead of "fall-throughs" in these kinds of non-trivial repetitions (and in the case of nesting, labelled loops with labelled continues). That can more easily catch a case like this where it's hard to tell what portion of the logic needs to repeat. Here's an example from pubsub streaming https://github.com/standard-ai/ya-gcp/blob/d257f5684477bea2a59314e0515a20ce72c20a3f/src/pubsub/streaming_subscription.rs#L523-L538

Reply:
Oh, good catch! As you can see, I haven't figured out a way to test the retry path. I thought about making a `tonic::transport::Channel` that injects errors somehow, but it didn't seem super easy. Do you know of a good way?

Reply:
I don't know an _easy_ way. The most robust way i can think of is to use the proto service schema to generate a stub bigtable server, and then mock messages+retriable errors. That's of course a lot of setup, but provides good abstraction without touching the client.

There could be easier solutions, I'm not sure